### ISSUE TITLE
Move torchvision.utils.save_image -> torchvision.io.save_image (or introduce this function from scratch) + support of batches / parallelization

### ISSUE THREAD

**Issue by vadimkantorov** [pytorch/vision#5461]
### 🚀 The feature

I think the `io` namespace makes more sense, and it's where `torchvision.io.read_image` is located...

### Motivation, pitch

N/A

**Comment by datumbox** [comment]
Yes, I agree. Not sure why it was added on utils but I agree it makes sense to be within io. As this is BC, we should rollout our new deprecation policy and then investigate impact and whether it's worth moving it.

**Comment by oke-aditya** [comment]
`utils.save_image` is a very old function (since 0,4 I suppose much before `io.write_jpeg`), i had pointed this to @NicolasHug too in offline discussion of deprecating it use `io.write_jpeg`. or `io.write_png`

They small difference is utils.,save_image saves an image tensor. Whereas io does over PIIL images. (Anyways the image tensor can be converted to PIL with transforms.toPILImage)

**Comment by NicolasHug** [comment]
I don't think we should move `save_image`, in fact I think we should deprecate it. It's not working as users would expect it to work. It adds padding pixels by default, and it relies on `make_grid` which has a bad API. It's causing confusion to users, see e.g. https://github.com/pytorch/vision/issues/5211 and my reply there.

**Comment by vadimkantorov** [comment]
I don't mind creating of a completely new function with this functionality, and it could redirect the call to write_jpg / write_png accordingly to its arguments / file extension, just like opencv does. A usefsul feature could be supporting batches of images to parallelize image encoding and waiting for file system to complete saving.

**Comment by ivan-gorin** [comment]
I also think `save_image` should be deprecated, it assumes the input is in `[0, 1]` range and doesn't provide any indication that `[0, 255]` integer representation is not supported. This is also not explained in the documentation.

**Comment by vadimkantorov** [comment]
Also, batched/parallelized read_image api could save some simple cases where otherwise a full dataset/dataloader would be needed (if practical enough to implement it).

**Comment by ankitmaurya001** [comment]
I agree here for deprecating save_image.
Any ideas on how to save a uint64 tensor in the range[0,255] as an image?

**Comment by vadimkantorov** [comment]
> Also, batched/parallelized read_image api could save some simple cases where otherwise a full dataset/dataloader would be needed (if practical enough to implement it).

One in-the-wild usecase for a simple batch image reading function: https://github.com/pytorch/pytorch/issues/88978

**Comment by foreverseer** [comment]
It is rather strange for the user that functions with similar functions are placed in two different modules

**Comment by NicolasHug** [comment]
Sorry, moving `save_image` probably isn't going to happen, so I'll close this issue.

We acknowledge that the situation is not ideal and confusing, but moving it would still require a deprecation / disruption, and as explained above in https://github.com/pytorch/vision/issues/5461#issuecomment-1048612641 I don't think `save_image` should exist in the first place. Try using the better-supported alternatives detailed in https://github.com/pytorch/vision/issues/5211#issuecomment-1016255934

**Comment by vadimkantorov** [comment]
@NicolasHug this issue also proposes parallelized/batch image saving. Does `write_jpg` already support this feature or is it also wontfix?

**Comment by NicolasHug** [comment]
Feel free to open a separate an issue for that. We can clarify what parallelism means there.

(I'm not particularly sold on the "batch saving" idea considering it's really just a call to `torch.cat()`)

**Comment by vadimkantorov** [comment]
The idea is to have the I/O and encoding/decoding itself being done in parallel.

**Comment by vadimkantorov** [comment]
We also discussed in the issue creating of a new dispatching function the the proper torchvision.ok namespace (e.g. torchvision.io.write_image), so this issue was not just about moving the existing function (deemed obsolete and should-be-deprecated) to the proper namespace

**Comment by NicolasHug** [comment]
> done in parallel

multithreading? multi-processing?

Neither is going to help much when using the `DataLoader(num_workers=N)` because multi-threading is set to 1 and multiprocessing is already enabled. Apart from during training, what situations would "parallelism" be critical here?

**Comment by vadimkantorov** [comment]
You are right, inside DataLoader context this is not adding anything useful.

But this is useful in other contexts, e.g. when generating visualizations of many images or processing a directory of images as a single batch. This can simplify the code in simple scenarios, and this I/O can be paralellized. So yes, some simple python multiprocessing (or better OpenMP in C++ if C++ libtorch is still being maintained)

**Comment by vadimkantorov** [comment]
Okay, I will create new separate issues for these ideas

**Comment by vadimkantorov** [comment]
I consider this a matter of idioms / saving users' time reimplementing the common things / having tested utils.

For the visualization case, it's often lazy to be measuring/profiling plots (often hundreds of them) encoding to jpeg and base64, as this is often a scratch/"write only code to be thrown away". So I always kept them serial, but I suspect that image encoding and base64 encoding actually takes half of the time of running these scripts. These scripts end up running many times and each execution can easily start taking dozens of seconds, so at the end takes a lot of wasted time

For the "reading a directory of images for inference/testing" - I think also yes, parallelization can help and offset a lot of CPU-bound decoding and overlapped I/O and is definitely a simpler API than spinning a full DataLoader for this

**Comment by NicolasHug** [comment]
I highly recommend you to consider `joblib` @vadimkantorov . I think it does what you need (and more) with 0.01% of the effort it would take to implement a parallel "save_images()" util. Knowing the joblib devs, it's been tested to death and it's just as reliable as anything coming out of pytorch.
